--- compound_scorer.py	2019-09-12 12:39:42.842725942 +0200
+++ /home/bjorn/miniconda2/lib/python2.7/site-packages/genmod/score_variants/compound_scorer.py	2019-09-12 08:29:05.600059286 +0200
@@ -141,18 +141,18 @@
                     
                     logger.debug("Checking compounds for family {0}".format(
                         compound_family_id))
-                    
+
                     #Loop through compounds to check if they are only low scored
                     for compound_id in compound_list:
                         compound_rank_score = rank_scores[compound_id]
-                        if compound_rank_score > 9:
+                        if compound_rank_score > 12:
                             only_low = False
                     logger.debug("Setting only_low to {0}".format(only_low))
                     
                     if (correct_score and only_low):
                         logger.debug("correcting rank score for {0}".format(
                             variant_id))
-                        current_rank_score -= 6
+                        current_rank_score -= 8
                         
                     for compound_id in compound_list:
                         logger.debug("Checking compound {0}".format(compound_id))
