| Tool                   | Version      | Type        | URL                                                    | External contact                                       | Internal contact      | Short description                                                                         |
| ---------------------- | ------------ | ----------- | ------------------------------------------------------ | ------------------------------------------------------ | --------------------- | ----------------------------------------------------------------------------------------- |
| annotsv                | 2.3          | Open-Source | https://github.com/lgmgeo/AnnotSV                      | https://github.com/lgmgeo/AnnotSV/issues               | Viktor Henmyr         | Annotation and Ranking of Structural variants                                             |
| bcftools               | 1.9          | Open-Source | https://github.com/samtools/bcftools                   | https://github.com/samtools/bcftools/issues            | Jakob Willforss       | Variant calling and manipulating files in VCF/BCF format                                  |
| bedtools               | 2.29.0       | Open-Source | https://bedtools.readthedocs.io                        | https://github.com/arq5x/bedtools2/issues              | Sailendra Pradhananga | Tools for genome arithmetic                                                               |
| bwa                    | 0.7.17-r1188 | Open-Source | https://github.com/lh3/bwa                             | https://github.com/lh3/bwa/issues                      | Markus Johansson      | Aligner for short reads                                                                   |
| cadd                   | 1.6          | Open-Source | https://github.com/kircherlab/CADD-scripts             | https://github.com/kircherlab/CADD-scripts/issues      | Jakob Willforss       | Score the deleteriousness of SNV and indels in human genome                               |
| cnvkit                 | 0.9.6        | Open-Source | https://github.com/etal/cnvkit                         | https://github.com/etal/cnvkit/issues                  | Ram Sai Nanduri       | Copy number variant detection from targeted DNA sequencing                                |
| delly                  | 0.8.3        | Open-Source | https://github.com/dellytools/delly                    | https://github.com/dellytools/delly/issues             | Ryan Kennedy          |                                                                                           |
| eklipse                | 1.8          | Open-Source | https://github.com/dooguypapua/eKLIPse                 | https://github.com/dooguypapua/eKLIPse/issues          | Paul Piccinelli       | Detection and quantification of mitochondrial DNA deletions                               |
| expansionhunter        | 4.0.2        | Open-Source | https://github.com/Illumina/ExpansionHunter            | https://github.com/Illumina/ExpansionHunter/issues     | Paul Piccinelli       | Estimate repeat sizes                                                                     |
| fastp                  | 0.23.4       | Open-Source | https://github.com/OpenGene/fastp                      | https://github.com/OpenGene/fastp/issues               | Sailendra Pradhananga | FASTQ preprocessor                                                                        |
| freebayes              | 1.3.2        | Open-Source | https://github.com/freebayes/freebayes                 | https://github.com/freebayes/freebayes/issues          | Ram Sai Nanduri       | Bayesian haplotype-based genetic polymorphism discovery and genotyping                    |
| gatk                   | 4.1.9.0      | Open-Source | https://gatk.broadinstitute.org/hc/en-us               | https://github.com/broadinstitute/gatk/issues          | Viktor Henmyr         | Wide set of tools for genomic analysis                                                    |
| genmod                 | 3.7.4        | Open-Source | https://github.com/Clinical-Genomics/genmod            | https://github.com/Clinical-Genomics/genmod/issues     | Viktor Henmyr         | Annotate models of genetic inheritance patterns in variant files                          |
| haplogrep              | 2.2.8        | Open-Source | https://github.com/seppinho/haplogrep-cmd              | https://github.com/seppinho/haplogrep-cmd/issues       | Paul Piccinelli       | mtDNA haplogroup classification. Supporting rCRS and RSRS.                                |
| hmtnote                | 0.7.2        | Open-Source | https://github.com/robertopreste/HmtNote               | https://github.com/robertopreste/HmtNote/issues        | Paul Piccinelli       | Human mitochondrial variants annotation using HmtVar                                      |
| madeleine              | 2.0          | Open-Source | https://github.com/piratical/Madeline_2.0_PDE          | https://github.com/piratical/Madeline_2.0_PDE/issues   | Jakob Willforss       | Pedigree drawing program                                                                  |
| manta                  | 1.6.0        | Open-Source | https://github.com/Illumina/manta                      | https://github.com/Illumina/manta/issues               | Sailendra Pradhananga | Structural variant and indel caller for mapped sequencing data                            |
| melt                   | 2.2.2        | Open-Source | https://melt.igs.umaryland.edu                         | https://melt.igs.umaryland.edu                         | Viktor Henmyr         | Mobile element locator tool                                                               |
| montage                | 1.3.35       | Open-Source | https://github.com/CAG-CNV/MONTAGE                     | https://github.com/CAG-CNV/MONTAGE/issues              | Alexander Koc         | Mosaic CNV detection tool                                                                 |
| ped-parser             | 1.6.3        | Open-Source | https://github.com/moonso/ped_parser                   | https://github.com/moonso/ped_parser/issues            | Jakob Willforss       | Parse pedigree files                                                                      |
| peddy                  | 0.4.3        | Open-Source | https://github.com/brentp/peddy                        | https://github.com/brentp/peddy/issues                 | Jakob Willforss       | genotype :: ped correspondence check, ancestry check, sex check. directly, quickly on VCF |
| reviewer               | 0.2.1        | Open-Source | https://github.com/Illumina/REViewer                   | https://github.com/Illumina/REViewer/issues            | Paul Piccinelli       | Visualize alignments of reads in regions containing tandem repeats                        |
| sambamba               | 0.7.1        | Open-Source | https://lomereiter.github.io/sambamba/                 | https://github.com/biod/sambamba/issues                | Ram Sai Nanduri       | Tools for working with SAM/BAM data                                                       |
| samtools               | 1.9          | Open-Source | http://www.htslib.org/                                 | https://github.com/samtools/samtools/issues            | Markus Johansson      | Tools for manipulating next-generation sequencing data                                    |
| sentieon               | 202112       | Commercial  | https://www.sentieon.com/                              | https://www.sentieon.com/                              | Viktor Henmyr         | Suite of bioinformatics analysis tools                                                    |
| smn-copy-number-caller | 1.1.2        | Open-Source | https://github.com/Illumina/SMNCopyNumberCaller        | https://github.com/Illumina/SMNCopyNumberCaller/issues | Alexander Koc         | Copy number caller for SMN1 and SMN2                                                      |
| stranger               | 0.8.0        | Open-Source | https://github.com/Clinical-Genomics/stranger          | https://github.com/Clinical-Genomics/stranger/issues   | Paul Piccinelli       | Annotate outfiles from ExpansionHunter with the pathologic implications of the repeat     |
| svdb                   | 2.2.0        | Open-Source | https://github.com/J35P312/SVDB                        | https://github.com/J35P312/SVDB/issues                 | Alexander Koc         | Structural variant database software                                                      |
| tabix                  | 1.9          | Open-Source | http://www.htslib.org/doc/tabix.html                   |                                                        | Alexander Koc         |                                                                                           |
| tiddit                 | 2.8.1        | Open-Source | https://github.com/SciLifeLab/TIDDIT                   | https://github.com/SciLifeLab/TIDDIT/issues            | Jakob Willforss       | TIDDIT - structural variant calling                                                       |
| upd                    | 0.1          | Open-Source | https://github.com/bjhall/upd                          | https://github.com/bjhall/upd/issues                   | Viktor Henmyr         | Basic UPD caller                                                                          |
| vcfanno                | 0.3.2        | Open-Source | https://github.com/brentp/vcfanno                      | https://github.com/brentp/vcfanno/issues               | Sailendra Pradhananga | Annotate a VCF with other VCFs/BEDs/tabixed files                                         |
| vcflib                 | 1.0.9        | Open-Source | https://github.com/vcflib/vcflib                       | https://github.com/vcflib/vcflib/issues                | Alexander Koc         | Parse and manipulate VCF files with python and zig bindings                               |
| vep                    | 103.1        | Open-Source | https://www.ensembl.org/info/docs/tools/vep/index.html | https://github.com/Ensembl/ensembl-vep/issues          | Ram Sai Nanduri       | Predict functional effects of genomic variants                                            |
